src/main/java/reposense/model/RepoLocation.java

```diff
@@ -53,31 +54,108 @@ public String getOrganization() {
         return organization;
     }
 
+    public Optional<String> getParsedBranch() {
+        return parsedBranch;
+    }
+
+    /**
+     * Given a String representing a repo location, returns
+     * an array containing the following details of the repository (in order):
+     * { location, repository name, organisation name, branch name (if any) }
+     *
+     * @param location a repository location, which is a file path or URL with the branch
+     *                 name optionally appended
+     * @throws InvalidLocationException if the repo location is an invalid path or an invalid URL
+     */
+    private static String[] parse(String location) throws InvalidLocationException {
+        String[] parsedInfo = tryParsingAsRepoUrl(location);
+        if (parsedInfo != null) {
+            return parsedInfo;
+        }
+
+        parsedInfo = tryParsingAsPath(location);
+        if (parsedInfo != null) {
+            return parsedInfo;
+        }
+
+        parsedInfo = tryParsingAsBranchUrl(location);
+        if (parsedInfo != null) {
+            return parsedInfo;
+        }
+
+        throw new InvalidLocationException("The given location is invalid");
+    }
+
+    /**
+     * Parses a given repo URL and returns an array containing the following info:
+     * { url, repository name, organisation name, branch name (if any }
+     *
+     * @return null if the given String is an invalid URL
+     */
+    private static String[] tryParsingAsRepoUrl(String location) {
```

Comment:
I'm thinking whether if u can merge the methods `tryParsing...` as one as they have a lot of repeated logic.

Reply:
I've merged the logic responsible for parsing both types of URLs into a single method called
`tryParsingAsUrl`. 

The two types of URLs are 
(1) URLs to a branch on GitHub, (2) URLs to the `.git` folder on GitHub